store/postgres/src/detail.rs

```diff
@@ -360,3 +373,95 @@ pub fn deployment_entity(
 
     SubgraphDeploymentEntity::try_from(StoredDeploymentEntity(detail, manifest))
 }
+
+#[derive(Queryable, Identifiable, Insertable)]
+#[table_name = "graph_node_versions"]
+pub struct GraphNodeVersion {
+    pub id: i32,
+    pub git_commit_hash: String,
+    pub git_repository_dirty: bool,
+    pub crate_version: String,
+    pub major: i32,
+    pub minor: i32,
+    pub patch: i32,
+    pub pre_release: String,
+    pub rustc_version: String,
+    pub rustc_host: String,
+    pub rustc_channel: String,
+}
+
+impl GraphNodeVersion {
+    pub(crate) fn create_or_get(conn: &PgConnection) -> anyhow::Result<i32> {
+        let git_repository_path =
+            GraphNodeVersion::find_git_repository_path(std::path::Path::new(CARGO_MANIFEST_DIR))?;
+        let git_repository = Repository::open(&git_repository_path)?;
+        let last_commit = GraphNodeVersion::find_last_commit(&git_repository)?;
+        let git_commit_hash = last_commit.id().to_string();
+        let git_repository_dirty = GraphNodeVersion::is_repository_dirty(&git_repository)?;
+        let crate_version = CARGO_PKG_VERSION.to_string();
+        let pre_release = CARGO_PKG_VERSION_PRE.to_string();
+        let rustc = rustc_version::version_meta()?;
+        let rustc_channel = format!("{:?}", rustc.channel);
+
+        // Cargo won't run if those values are not integers, so it is safe to unwrap here.
+        let major: i32 = CARGO_PKG_VERSION_MAJOR.parse().unwrap();
+        let minor: i32 = CARGO_PKG_VERSION_MINOR.parse().unwrap();
+        let patch: i32 = CARGO_PKG_VERSION_PATCH.parse().unwrap();
+
+        // upsert
+        let graph_node_version_id = {
+            use graph_node_versions::dsl as g;
+            diesel::insert_into(g::graph_node_versions)
+                .values((
+                    g::git_commit_hash.eq(&git_commit_hash),
+                    g::git_repository_dirty.eq(git_repository_dirty),
+                    g::crate_version.eq(&crate_version),
+                    g::major.eq(&major),
+                    g::minor.eq(&minor),
+                    g::patch.eq(&patch),
+                    g::pre_release.eq(&pre_release),
+                    g::rustc_version.eq(&rustc.semver.to_string()),
+                    g::rustc_host.eq(&rustc.host.to_string()),
+                    g::rustc_channel.eq(&rustc_channel),
+                ))
+                .on_conflict(diesel::pg::upsert::on_constraint(
+                    "unique_graph_node_versions",
+                ))
+                .do_update()
```

Comment:
With the changes to constraints, it's better to make this `on_conflict_do_nothing` and then explicitly query for the version.

Reply:
Since the version info will be available as a `'static`, I believe we'll need only to issue an `INSERT` command, without the need for a `SELECT`, right?
If so, I agree that it could have `on_conflict_do_nothing`.